cli/command/image/build.go

```diff
@@ -528,57 +363,122 @@ func rewriteDockerfileFrom(ctx context.Context, dockerfile io.Reader, translator
-// new Dockerfile. Returns a new tar archive stream with the replaced
-// Dockerfile.
-func replaceDockerfileTarWrapper(ctx context.Context, inputTarStream io.ReadCloser, dockerfileName string, translator translatorFunc, resolvedTags *[]*resolvedTag) io.ReadCloser {
-	pipeReader, pipeWriter := io.Pipe()
-	go func() {
-		tarReader := tar.NewReader(inputTarStream)
-		tarWriter := tar.NewWriter(pipeWriter)
-
-		defer inputTarStream.Close()
-
-		for {
-			hdr, err := tarReader.Next()
-			if err == io.EOF {
-				// Signals end of archive.
-				tarWriter.Close()
-				pipeWriter.Close()
-				return
-			}
-			if err != nil {
-				pipeWriter.CloseWithError(err)
-				return
-			}
+func rewriteDockerfileForContentTrust(buildCtx io.ReadCloser, dockerfileName string, translator translatorFunc) (io.ReadCloser, []*resolvedTag) {
+	var resolvedTags []*resolvedTag
+	return archive.ReplaceFileTarWrapper(buildCtx, map[string]archive.TarModifierFunc{
+		dockerfileName: func(_ string, hdr *tar.Header, content io.Reader) (*tar.Header, []byte, error) {
+			var newDockerfile []byte
+			var err error
+			newDockerfile, resolvedTags, err = rewriteDockerfileFrom(content, translator)
+			hdr.Size = int64(len(newDockerfile))
+			return hdr, newDockerfile, err
+		},
+	}), resolvedTags
+}
 
-			content := io.Reader(tarReader)
-			if hdr.Name == dockerfileName {
-				// This entry is the Dockerfile. Since the tar archive was
-				// generated from a directory on the local filesystem, the
-				// Dockerfile will only appear once in the archive.
-				var newDockerfile []byte
-				newDockerfile, *resolvedTags, err = rewriteDockerfileFrom(ctx, content, translator)
-				if err != nil {
-					pipeWriter.CloseWithError(err)
-					return
-				}
-				hdr.Size = int64(len(newDockerfile))
-				content = bytes.NewBuffer(newDockerfile)
-			}
+type buildBuffer struct {
+	quiet    bool
+	stderr   io.Writer
+	stdout   io.Writer
+	output   io.Writer
+	progress io.Writer
+}
 
-			if err := tarWriter.WriteHeader(hdr); err != nil {
-				pipeWriter.CloseWithError(err)
-				return
-			}
+func newBuildBuffer(out io.Writer, stderr io.Writer, quiet bool) *buildBuffer {
+	stdout := out
+	progress := out
+	if quiet {
+		out = bytes.NewBuffer(nil)
+		progress = bytes.NewBuffer(nil)
+	}
+	return &buildBuffer{
+		output:   out,
+		progress: progress,
+		stdout:   stdout,
+		quiet:    quiet,
+		stderr:   stderr,
+	}
+}
+
+// PrintErrorIfQuiet to the stderr stream. If quiet is not set then the error
+// was printed elsewhere, so do nothing.
+func (bb *buildBuffer) PrintErrorIfQuiet() {
+	if bb.quiet {
+		fmt.Fprintf(bb.stderr, "%s%s", bb.progress, bb.output)
+	}
+}
 
-			if _, err := io.Copy(tarWriter, content); err != nil {
-				pipeWriter.CloseWithError(err)
-				return
+// PrintImageIDIfQuiet to stdout. If quiet is not set then the image ID is
+// printed elsewhere, so do nothing.
+func (bb *buildBuffer) PrintImageIDIfQuiet() {
+	if bb.quiet {
+		fmt.Fprintf(bb.stdout, "%s", bb.output)
```

Comment:
Fair, We should cast it to `bytes.Buffer` first, and panic if it's the wrong type.

Since this is now encapsulated, it's easy to verify that the writer is the correct type when `quiet`.